server/auth.go

```diff
@@ -0,0 +1,158 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/prometheus/common/log"
+)
+
+var (
+	ErrUnauthorized = errors.New("jwtauth: token is unauthorized")
+	ErrExpired      = errors.New("jwtauth: token is expired")
+	ErrNBFInvalid   = errors.New("jwtauth: token nbf validation failed")
+	ErrIATInvalid   = errors.New("jwtauth: token iat validation failed")
+	ErrNoTokenFound = errors.New("jwtauth: no token found")
+	ErrAlgoInvalid  = errors.New("jwtauth: algorithm mismatch")
+)
+
+type JWTAuth struct {
+	signKey   interface{}
+	verifyKey interface{}
+	signer    jwt.SigningMethod
+	parser    *jwt.Parser
+}
+
+var (
+	TokenCtxKey = &contextKey{"Token"}
+	ErrorCtxKey = &contextKey{"Error"}
+)
+
+var tokenAuth *JWTAuth
+
+func InitAuth(secretKey []byte) {
+	tokenAuth = &JWTAuth{
+		signKey:   secretKey,
+		verifyKey: nil,
+		signer:    jwt.GetSigningMethod("HS256"),
+		parser:    &jwt.Parser{},
+	}
+}
+
+func (ja *JWTAuth) Encode(claims jwt.Claims) (t *jwt.Token, tokenString string, err error) {
+	t = jwt.New(ja.signer)
+	t.Claims = claims
+	tokenString, err = t.SignedString(ja.signKey)
+	t.Raw = tokenString
+	return
+}
+
+func (ja *JWTAuth) Decode(tokenString string) (t *jwt.Token, err error) {
+	t, err = ja.parser.Parse(tokenString, ja.keyFunc)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func (ja *JWTAuth) keyFunc(t *jwt.Token) (interface{}, error) {
+	if ja.verifyKey != nil {
+		return ja.verifyKey, nil
+	} else {
+		return ja.signKey, nil
+	}
+}
+
+func NewAuthContext(ctx context.Context, t *jwt.Token, err error) context.Context {
+	ctx = context.WithValue(ctx, TokenCtxKey, t)
+	ctx = context.WithValue(ctx, ErrorCtxKey, err)
+	return ctx
+}
+
+func FromAuthContext(ctx context.Context) (*jwt.Token, jwt.MapClaims, error) {
+	token, _ := ctx.Value(TokenCtxKey).(*jwt.Token)
+
+	var claims jwt.MapClaims
+	if token != nil {
+		if tokenClaims, ok := token.Claims.(jwt.MapClaims); ok {
+			claims = tokenClaims
+		} else {
+			panic(fmt.Sprintf("jwtauth: unknown type of Claims: %T", token.Claims))
+		}
+	} else {
+		claims = jwt.MapClaims{}
+	}
+
+	err, _ := ctx.Value(ErrorCtxKey).(error)
+
+	return token, claims, err
+}
+
+// JWTFromFromCookie tries to retrieve the jwt.MapClaims from a cookie named
+// "jwt".
+func JWTFromFromCookie(r *http.Request) (jwt.MapClaims, error) {
```

Comment:
```suggestion
func JWTFromCookie(r *http.Request) (jwt.MapClaims, error) {
```

Change usage also please

Reply:
I would vote for `getJwtFromCookie`